File: link_checker/checker.py
"""Link status checking and redirect tracing module."""

import logging
import time
from dataclasses import dataclass, field
from typing import Optional

import requests

from .utils import format_redirect_chain, get_domain, is_canonical_redirect

logger = logging.getLogger(__name__)

MAX_REDIRECTS = 10


@dataclass
class LinkStatus:
    """Result of checking a single link."""
    url: str
    status_code: int
    is_redirect: bool = False
    redirect_chain: list[tuple[int, str]] = field(default_factory=list)
    final_url: str = ''
    is_loop: bool = False
    is_canonical_redirect: bool = False
    error: str = ''
    retries: int = 0  # Number of retries needed for transient errors
    
    @property
    def redirect_chain_formatted(self) -> str:
        """Get the redirect chain as a formatted string."""
        return format_redirect_chain(self.redirect_chain)


class LinkChecker:
    """Checks link status and traces redirect chains."""
    
    # Status codes that trigger retry
    RETRY_STATUS_CODES = {502, 503, 504}
    # Default retry settings
    DEFAULT_MAX_RETRIES = 3
    DEFAULT_RETRY_DELAY = 1.0  # seconds
    DEFAULT_RETRY_BACKOFF = 2.0  # multiplier
    
    def __init__(
        self,
        user_agent: str = 'LinkCanary/1.0',
        timeout: int = 10,
        delay: float = 0.1,
        max_retries: int = 3,
        retry_delay: float = 1.0,
        retry_backoff: float = 2.0,
        auth_user: Optional[str] = None,
        auth_pass: Optional[str] = None,
        headers: Optional[dict] = None,
        cookies: Optional[dict] = None,
    ):
        self.user_agent = user_agent
        self.timeout = timeout
        self.delay = delay
        self.max_retries = max_retries
        self.retry_delay = retry_delay
        self.retry_backoff = retry_backoff
        self.auth_user = auth_user
        self.auth_pass = auth_pass
        self.custom_headers = headers or {}
        self.cookies = cookies
        
        self.session = requests.Session()
        self.session.headers.update({
            'User-Agent': self.user_agent,
            'Accept': '*/*',
        })
        
        # Add custom headers
        if self.custom_headers:
            self.session.headers.update(self.custom_headers)
        
        # Add basic auth
        if self.auth_user and self.auth_pass:
            self.session.auth = (self.auth_user, self.auth_pass)
        
        # Add cookies
        if self.cookies:
            self.session.cookies.update(self.cookies)
        
        self._cache: dict[str, LinkStatus] = {}
        self._head_blacklist: set[str] = set()
        self._host_delays: dict[str, float] = {}
        self._last_request_time: dict[str, float] = {}
        self._retry_stats: dict[str, int] = {}  # Track retry counts per URL
    
    def _get_host_delay(self, url: str) -> float:
        """Get the current delay for a host."""
        host = get_domain(url)
        return self._host_delays.get(host, self.delay)
    
    def _increase_host_delay(self, url: str):
        """Double the delay for a host (rate limit backoff)."""
        host = get_domain(url)
        current = self._host_delays.get(host, self.delay)
        self._host_delays[host] = min(current * 2, 30.0)
        logger.info(f"Increased delay for {host} to {self._host_delays[host]}s")
    
    def _rate_limit(self, url: str):
        """Apply rate limiting for a specific host."""
        host = get_domain(url)
        last_time = self._last_request_time.get(host, 0)
        delay = self._get_host_delay(url)
        
        elapsed = time.time() - last_time
        if elapsed < delay:
            time.sleep(delay - elapsed)
        
        self._last_request_time[host] = time.time()
    
    def _should_use_get(self, url: str) -> bool:
        """Check if we should skip HEAD and use GET directly."""
        host = get_domain(url)
        return host in self._head_blacklist
    
    def _add_to_head_blacklist(self, url: str):
        """Add a host to the HEAD request blacklist."""
        host = get_domain(url)
        if host not in self._head_blacklist:
            self._head_blacklist.add(host)
            logger.debug(f"Added {host} to HEAD blacklist")
    
    def _make_request(
        self,
        url: str,
        method: str = 'HEAD',
        allow_redirects: bool = False,
    ) -> Optional[requests.Response]:
        """Make an HTTP request with error handling."""
        self._rate_limit(url)
        
        try:
            if method == 'HEAD':
                response = self.session.head(
                    url,
                    timeout=self.timeout,
                    allow_redirects=allow_redirects,
                )
            else:
                response = self.session.get(
                    url,
                    timeout=self.timeout,
                    allow_redirects=allow_redirects,
                    stream=True,
                )
            return response
        except requests.Timeout:
            return None
        except requests.RequestException:
            return None
    
    def _make_request_with_retry(
        self,
        url: str,
        method: str = 'HEAD',
        allow_redirects: bool = False,
    ) -> tuple[Optional[requests.Response], int]:
        """
        Make an HTTP request with retry logic for transient errors.
        
        Retries on:
        - 502 Bad Gateway
        - 503 Service Unavailable
        - 504 Gateway Timeout
        
        Uses exponential backoff between retries.
        
        Returns:
            Tuple of (response, retry_count)
        """
        retry_count = 0
        current_delay = self.retry_delay
        
        while retry_count <= self.max_retries:
            response = self._make_request(url, method, allow_redirects)
            
            # No response (timeout/connection error) - might be transient
            if response is None:
                if retry_count < self.max_retries:
                    retry_count += 1
                    logger.debug(f"Retrying {url} (attempt {retry_count}/{self.max_retries}) - no response")
                    time.sleep(current_delay)
                    current_delay *= self.retry_backoff
                    continue
                return None, retry_count
            
            # Success or non-retryable status
            if response.status_code not in self.RETRY_STATUS_CODES:
                return response, retry_count
            
            # Retryable server error (502/503/504)
            if retry_count < self.max_retries:
                retry_count += 1
                logger.debug(
                    f"Retrying {url} (attempt {retry_count}/{self.max_retries}) - "
                    f"status {response.status_code}"
                )
                time.sleep(current_delay)
                current_delay *= self.retry_backoff
            else:
                # Max retries reached, return the error response
                return response, retry_count
        
        return response, retry_count
    
    def _check_single_url(self, url: str) -> tuple[int, str, Optional[str], int]:
        """
        Check a single URL's status without following redirects.
        
        Returns:
            Tuple of (status_code, redirect_location or '', error_message or None, retry_count)
        """
        use_get = self._should_use_get(url)
        retry_count = 0
        
        if not use_get:
            response, retry_count = self._make_request_with_retry(url, 'HEAD', allow_redirects=False)
            
            if response is not None:
                if response.status_code in (403, 405, 501):
                    self._add_to_head_blacklist(url)
                    use_get = True
                else:
                    location = response.headers.get('Location', '')
                    return response.status_code, location, None, retry_count
        
        if use_get or response is None:
            response, retry_count = self._make_request_with_retry(url, 'GET', allow_redirects=False)
        
        if response is None:
            return 0, '', 'Request failed (timeout or connection error)', retry_count
        
        if response.status_code == 429:
            self._increase_host_delay(url)
            time.sleep(self._get_host_delay(url))
            response, retry_count = self._make_request_with_retry(url, 'GET', allow_redirects=False)
            
            if response is None or response.status_code == 429:
                self._increase_host_delay(url)
                time.sleep(self._get_host_delay(url))
                response, retry_count = self._make_request_with_retry(url, 'GET', allow_redirects=False)
                
                if response is None or response.status_code == 429:
                    return 429, '', 'Rate limited after retries', retry_count
        
        location = response.headers.get('Location', '')
        return response.status_code, location, None, retry_count
    
    def check_link(self, url: str) -> LinkStatus:
        """
        Check a link's status, following redirects and detecting issues.
        
        Args:
            url: The URL to check
        
        Returns:
            LinkStatus object with all details
        """
        if url in self._cache:
            return self._cache[url]
        
        chain: list[tuple[int, str]] = []
        visited: set[str] = set()
        current_url = url
        is_loop = False
        error = ''
        total_retries = 0
        
        for _ in range(MAX_REDIRECTS + 1):
            if current_url in visited:
                is_loop = True
                break
            
            visited.add(current_url)
            status_code, location, req_error, retry_count = self._check_single_url(current_url)
            total_retries += retry_count
            
            if req_error:
                error = req_error
                chain.append((0, current_url))
                break
            
            chain.append((status_code, current_url))
            
            if status_code < 300 or status_code >= 400:
                break
            
            if not location:
                break
            
            if not location.startswith('http'):
                from urllib.parse import urljoin
                location = urljoin(current_url, location)
            
            current_url = location
        
        if chain:
            final_status = chain[-1][0]
            final_url = chain[-1][1]
        else:
            final_status = 0
            final_url = url
        
        is_redirect = len(chain) > 1
        
        is_canonical = False
        if is_redirect and len(chain) == 2 and not is_loop:
            is_canonical = is_canonical_redirect(url, final_url)
        
        result = LinkStatus(
            url=url,
            status_code=final_status,
            is_redirect=is_redirect,
            redirect_chain=chain,
            final_url=final_url,
            is_loop=is_loop,
            is_canonical_redirect=is_canonical,
            error=error,
            retries=total_retries,
        )
        
        self._cache[url] = result
        return result
    
    def check_links(
        self,
        urls: list[str],
        progress_callback=None,
    ) -> dict[str, LinkStatus]:
        """
        Check multiple links.
        
        Args:
            urls: List of URLs to check
            progress_callback: Optional callback for progress updates
        
        Returns:
            Dictionary mapping URL to LinkStatus
        """
        results = {}
        
        for i, url in enumerate(urls):
            status = self.check_link(url)
            results[url] = status
            
            if progress_callback:
                progress_callback(i + 1, len(urls), url, status)
        
        return results
    
    def get_cache_stats(self) -> dict:
        """Get cache statistics."""
        total_retries = sum(status.retries for status in self._cache.values())
        urls_with_retries = sum(1 for status in self._cache.values() if status.retries > 0)
        
        return {
            'cached_urls': len(self._cache),
            'head_blacklisted_hosts': len(self._head_blacklist),
            'hosts_with_custom_delay': len(self._host_delays),
            'total_retries': total_retries,
            'urls_with_retries': urls_with_retries,
        }
    
    def close(self):
        """Close the session."""
        self.session.close()
    
    def __enter__(self):
        return self
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        self.close()
        return False

File: crawl-engine/checker.go
package main

import (
	"crypto/tls"
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"time"
)

// --- Link checker with HEAD/GET, redirect tracing, retry ---
// Port of link_checker/checker.py

const MaxRedirects = 10

var RetryStatusCodes = map[int]bool{502: true, 503: true, 504: true}

type LinkChecker struct {
	UserAgent     string
	Timeout       time.Duration
	Delay         time.Duration
	MaxRetries    int
	RetryDelay    time.Duration
	RetryBackoff  float64
	AuthUser      string
	AuthPass      string
	CustomHeaders map[string]string
	Cookies       map[string]string

	Client *http.Client

	cache           map[string]LinkStatus
	headBlacklist   map[string]bool
	hostDelays      map[string]time.Duration
	lastRequestTime map[string]time.Time
	retryStats      map[string]int
	mu              sync.Mutex
}

func NewLinkChecker(userAgent string, timeoutSec int, delaySec float64, maxRetries int, retryDelaySec, retryBackoff float64) *LinkChecker {
	lc := &LinkChecker{
		UserAgent:     userAgent,
		Timeout:       time.Duration(timeoutSec) * time.Second,
		Delay:         time.Duration(delaySec * float64(time.Second)),
		MaxRetries:    maxRetries,
		RetryDelay:    time.Duration(retryDelaySec * float64(time.Second)),
		RetryBackoff:  retryBackoff,
		CustomHeaders: make(map[string]string),
		Cookies:       make(map[string]string),
		Client: &http.Client{
			Timeout: time.Duration(timeoutSec) * time.Second,
			CheckRedirect: func(req *http.Request, via []*http.Request) error {
				return http.ErrUseLastResponse // Don't follow redirects automatically
			},
			Transport: &http.Transport{
				TLSClientConfig: &tls.Config{InsecureSkipVerify: false},
			},
		},
		cache:           make(map[string]LinkStatus),
		headBlacklist:   make(map[string]bool),
		hostDelays:      make(map[string]time.Duration),
		lastRequestTime: make(map[string]time.Time),
		retryStats:      make(map[string]int),
	}
	return lc
}

func (lc *LinkChecker) getHostDelay(rawURL string) time.Duration {
	host := GetDomain(rawURL)
	lc.mu.Lock()
	defer lc.mu.Unlock()
	if d, ok := lc.hostDelays[host]; ok {
		return d
	}
	return lc.Delay
}

func (lc *LinkChecker) increaseHostDelay(rawURL string) {
	host := GetDomain(rawURL)
	lc.mu.Lock()
	defer lc.mu.Unlock()
	current := lc.Delay
	if d, ok := lc.hostDelays[host]; ok {
		current = d
	}
	newDelay := current * 2
	if newDelay > 30*time.Second {
		newDelay = 30 * time.Second
	}
	lc.hostDelays[host] = newDelay
}

func (lc *LinkChecker) rateLimit(rawURL string) {
	host := GetDomain(rawURL)
	lc.mu.Lock()
	lastTime := lc.lastRequestTime[host]
	lc.mu.Unlock()
	delay := lc.getHostDelay(rawURL)
	elapsed := time.Since(lastTime)
	if elapsed < delay {
		time.Sleep(delay - elapsed)
	}
	lc.mu.Lock()
	lc.lastRequestTime[host] = time.Now()
	lc.mu.Unlock()
}

func (lc *LinkChecker) shouldUseGet(rawURL string) bool {
	host := GetDomain(rawURL)
	lc.mu.Lock()
	defer lc.mu.Unlock()
	return lc.headBlacklist[host]
}

func (lc *LinkChecker) addToHeadBlacklist(rawURL string) {
	host := GetDomain(rawURL)
	lc.mu.Lock()
	defer lc.mu.Unlock()
	lc.headBlacklist[host] = true
}

func (lc *LinkChecker) makeRequest(rawURL, method string) (*http.Response, float64) {
	lc.rateLimit(rawURL)

	var bodyReader *strings.Reader
	if method == "GET" {
		bodyReader = strings.NewReader("")
	}

	var req *http.Request
	var err error
	if method == "GET" && bodyReader != nil {
		req, err = http.NewRequest("GET", rawURL, bodyReader)
	} else {
		req, err = http.NewRequest(method, rawURL, nil)
	}
	if err != nil {
		return nil, 0
	}

	req.Header.Set("User-Agent", lc.UserAgent)
	req.Header.Set("Accept", "*/*")
	for k, v := range lc.CustomHeaders {
		req.Header.Set(k, v)
	}
	for k, v := range lc.Cookies {
		req.AddCookie(&http.Cookie{Name: k, Value: v})
	}
	if lc.AuthUser != "" && lc.AuthPass != "" {
		req.SetBasicAuth(lc.AuthUser, lc.AuthPass)
	}

	start := time.Now()
	resp, err := lc.Client.Do(req)
	elapsedMs := float64(time.Since(start).Microseconds()) / 1000.0
	if err != nil {
		return nil, 0
	}
	return resp, elapsedMs
}

func (lc *LinkChecker) makeRequestWithRetry(rawURL, method string) (*http.Response, int, float64) {
	retryCount := 0
	currentDelay := lc.RetryDelay

	for retryCount <= lc.MaxRetries {
		resp, elapsedMs := lc.makeRequest(rawURL, method)

		if resp == nil {
			if retryCount < lc.MaxRetries {
				retryCount++
				time.Sleep(currentDelay)
				currentDelay = time.Duration(float64(currentDelay) * lc.RetryBackoff)
				continue
			}
			return nil, retryCount, elapsedMs
		}

		if !RetryStatusCodes[resp.StatusCode] {
			return resp, retryCount, elapsedMs
		}

		if retryCount < lc.MaxRetries {
			retryCount++
			time.Sleep(currentDelay)
			currentDelay = time.Duration(float64(currentDelay) * lc.RetryBackoff)
		} else {
			return resp, retryCount, elapsedMs
		}
	}
	return nil, retryCount, 0
}

func (lc *LinkChecker) checkSingleURL(rawURL string) (int, string, string, int, float64) {
	useGet := lc.shouldUseGet(rawURL)
	var retryCount int
	var elapsedMs float64
	var resp *http.Response

	if !useGet {
		resp, retryCount, elapsedMs = lc.makeRequestWithRetry(rawURL, "HEAD")
		if resp != nil {
			if resp.StatusCode == 403 || resp.StatusCode == 405 || resp.StatusCode == 501 {
				lc.addToHeadBlacklist(rawURL)
				resp.Body.Close()
				useGet = true
			} else {
				location := resp.Header.Get("Location")
				status := resp.StatusCode
				resp.Body.Close()
				return status, location, "", retryCount, elapsedMs
			}
		}
	}

	if useGet || resp == nil {
		resp, retryCount, elapsedMs = lc.makeRequestWithRetry(rawURL, "GET")
	}

	if resp == nil {
		return 0, "", "Request failed (timeout or connection error)", retryCount, elapsedMs
	}

	if resp.StatusCode == 429 {
		lc.increaseHostDelay(rawURL)
		time.Sleep(lc.getHostDelay(rawURL))
		resp.Body.Close()
		resp, retryCount, elapsedMs = lc.makeRequestWithRetry(rawURL, "GET")

		if resp == nil || resp.StatusCode == 429 {
			lc.increaseHostDelay(rawURL)
			time.Sleep(lc.getHostDelay(rawURL))
			if resp != nil {
				resp.Body.Close()
			}
			resp, retryCount, elapsedMs = lc.makeRequestWithRetry(rawURL, "GET")

			if resp == nil || resp.StatusCode == 429 {
				if resp != nil {
					resp.Body.Close()
				}
				return 429, "", "Rate limited after retries", retryCount, elapsedMs
			}
		}
	}

	location := resp.Header.Get("Location")
	status := resp.StatusCode
	resp.Body.Close()
	return status, location, "", retryCount, elapsedMs
}

// CheckLink checks a URL's status, following redirects and detecting issues.
func (lc *LinkChecker) CheckLink(rawURL string) LinkStatus {
	lc.mu.Lock()
	if cached, ok := lc.cache[rawURL]; ok {
		lc.mu.Unlock()
		return cached
	}
	lc.mu.Unlock()

	var chain [][2]string
	visited := make(map[string]bool)
	currentURL := rawURL
	isLoop := false
	errMsg := ""
	totalRetries := 0
	var firstResponseTimeMs *float64

	for i := 0; i < MaxRedirects+1; i++ {
		if visited[currentURL] {
			isLoop = true
			break
		}
		visited[currentURL] = true

		statusCode, location, reqError, retryCount, elapsedMs := lc.checkSingleURL(currentURL)
		totalRetries += retryCount

		if firstResponseTimeMs == nil {
			firstResponseTimeMs = &elapsedMs
		}

		if reqError != "" {
			errMsg = reqError
			chain = append(chain, [2]string{"0", currentURL})
			break
		}

		chain = append(chain, [2]string{fmt.Sprintf("%d", statusCode), currentURL})

		if statusCode < 300 || statusCode >= 400 {
			break
		}
		if location == "" {
			break
		}
		if !strings.HasPrefix(location, "http") {
			location = ResolveRelativeURL(currentURL, location)
		}
		currentURL = location
	}

	var finalStatus int
	var finalURL string
	if len(chain) > 0 {
		fmt.Sscanf(chain[len(chain)-1][0], "%d", &finalStatus)
		finalURL = chain[len(chain)-1][1]
	} else {
		finalURL = rawURL
	}

	isRedirect := len(chain) > 1
	isCanonical := false
	if isRedirect && len(chain) == 2 && !isLoop {
		isCanonical = IsCanonicalRedirect(rawURL, finalURL)
	}

	result := LinkStatus{
		URL:              rawURL,
		StatusCode:       finalStatus,
		IsRedirect:       isRedirect,
		RedirectChain:    chain,
		FinalURL:         finalURL,
		IsLoop:           isLoop,
		IsCanonicalRedir: isCanonical,
		Error:            errMsg,
		Retries:          totalRetries,
	}
	if firstResponseTimeMs != nil {
		result.ResponseTimeMs = *firstResponseTimeMs
	}

	lc.mu.Lock()
	lc.cache[rawURL] = result
	lc.mu.Unlock()
	return result
}

// CheckLinks checks multiple URLs sequentially.
func (lc *LinkChecker) CheckLinks(urls []string) map[string]LinkStatus {
	results := make(map[string]LinkStatus)
	for _, u := range urls {
		results[u] = lc.CheckLink(u)
	}
	return results
}

// GetRetryStats returns retry statistics.
func (lc *LinkChecker) GetRetryStats() (int, int) {
	totalRetries := 0
	urlsWithRetries := 0
	lc.mu.Lock()
	defer lc.mu.Unlock()
	for _, status := range lc.cache {
		totalRetries += status.Retries
		if status.Retries > 0 {
			urlsWithRetries++
		}
	}
	return urlsWithRetries, totalRetries
}

// ToJSONStatus converts a LinkStatus to its JSON-serializable form.
func (ls *LinkStatus) ToJSONStatus() JSONLinkStatus {
	hops := make([]JSONRedirectHop, len(ls.RedirectChain))
	for i, hop := range ls.RedirectChain {
		var statusInt int
		fmt.Sscanf(hop[0], "%d", &statusInt)
		hops[i] = JSONRedirectHop{Status: statusInt, URL: hop[1]}
	}
	return JSONLinkStatus{
		URL:              ls.URL,
		StatusCode:       ls.StatusCode,
		IsRedirect:       ls.IsRedirect,
		RedirectChain:    hops,
		FinalURL:         ls.FinalURL,
		IsLoop:           ls.IsLoop,
		IsCanonicalRedir: ls.IsCanonicalRedir,
		Error:            ls.Error,
		Retries:          ls.Retries,
		ResponseTimeMs:   ls.ResponseTimeMs,
	}
}

// Ensure url.Parse is referenced (used indirectly via ResolveRelativeURL)
var _ = url.Parse
